src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -388,15 +472,25 @@ private bool ShouldInvalidateServer(Exception exception)
                         code = (ServerErrorCode)writeConcernError.GetValue("code", -1).ToInt32();
                         message = writeConcernError.GetValue("errmsg", null)?.AsString;
 
-                        if (IsNotMasterOrRecovering(code, message))
+                        if (IsStateChangeError(code, message))
                         {
-                            return true;
+                            return !IsStaleStateChangeError(commandException.Result, out responseTopologyVersion);
                         }
                     }
                 }
             }
 
+            responseTopologyVersion = null;
             return false;
+
+            bool IsStaleStateChangeError(BsonDocument response, out TopologyDescription? responseTopologyDescriptionVersion)
+            {
+                responseTopologyDescriptionVersion = TopologyDescription.FromMongoCommandResponse(response);
+                return description.TopologyVersion.IsFresherThanOrEqualToServerResponse( responseTopologyDescriptionVersion);
+                // We use FresherThanOrEqualTo because a state change should come with a new topology version
+                // This is equivalent to description.TopologyVersion.CompareFreshnessToServerResponse(responseTopologyDescriptionVersion) >= 0;
+            }
```

Comment:
From this latter part of the pseudo code in the [spec](https://github.com/mongodb/specifications/commit/7dd9c008be6edaf65368787d5e91c2ac3aeb95b8#diff-62b4da2e38e7d2a9c1a2d92d315895bbR1401-R1411)
```
  def isStaleError(topologyDescription, response):
        currentServer = topologyDescription.servers[server.address]
        currentGeneration = currentServer.pool.generation
        generation = get connection generation from response
        if generation < currentGeneration:
            # Stale generation number.
            return True

        currentTopologyVersion = currentServer.topologyVersion
        # True if the current error's topologyVersion is greater than the server's
        return compareTopologyVersion(currentTopologyVersion, response.get("topologyVersion")) >= 0
```

Note that we check for a stale connection generation before calling this. I am open to the idea of checking here as well (see earlier thread)

Reply:
Moving this comment to an non oudated line so it shows up in the diff. Resolving.